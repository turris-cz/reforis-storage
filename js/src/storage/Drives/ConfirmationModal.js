/*
 * Copyright (C) 2020-2024 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { Modal, ModalBody, ModalHeader, ModalFooter, Button } from "foris";
import PropTypes from "prop-types";

ConfirmationModal.propTypes = {
    shown: PropTypes.bool.isRequired,
    setShown: PropTypes.func.isRequired,
    onConfirm: PropTypes.func.isRequired,
    isFirstDrive: PropTypes.bool.isRequired,
};

const FIRST_DRIVE_MESSAGE = _(`
You are adding the first drive, then you'll get notification and you will be asked to reboot. On the reboot data will be
moved from old drive to the new one. This can take some time so your next reboot will take longer. 
`);

export default function ConfirmationModal({
    shown,
    setShown,
    onConfirm,
    isFirstDrive,
}) {
    const firstDriveMessage = isFirstDrive ? FIRST_DRIVE_MESSAGE : "";
    return (
        <Modal shown={shown} setShown={setShown}>
            <ModalHeader setShown={setShown} title={_("Warning!")} />
            <ModalBody>
                <p>
                    {_(`
Are you sure you want to proceed? Newly selected drive(s) will be formatted and you will loose all the data on it.
${firstDriveMessage}
Are you sure you want to continue?
                    `)}
                </p>
            </ModalBody>
            <ModalFooter>
                <Button
                    className="btn-secondary"
                    onClick={() => setShown(false)}
                >
                    {_("Cancel")}
                </Button>
                <Button className="btn-primary" onClick={onConfirm}>
                    {_("Continue")}
                </Button>
            </ModalFooter>
        </Modal>
    );
}
