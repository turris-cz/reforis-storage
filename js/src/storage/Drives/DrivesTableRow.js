/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useState } from "react";

import { CheckBox } from "foris";
import update from "immutability-helper";
import PropTypes from "prop-types";

DriveTableRow.propTypes = {
    drive: PropTypes.shape({
        dev: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired,
        fs: PropTypes.string.isRequired,
        uuid: PropTypes.string.isRequired,
    }).isRequired,
    isSelected: PropTypes.bool.isRequired,
    setSelectedDrives: PropTypes.func.isRequired,
    active: PropTypes.bool.isRequired,
    disabled: PropTypes.bool.isRequired,
};

export default function DriveTableRow({
    drive,
    isSelected,
    setSelectedDrives,
    active,
    disabled,
}) {
    const [isActiveAndSelected, setIsActiveAndSelected] = useState(active);

    function select() {
        setSelectedDrives((selectedDrives) => {
            const index = selectedDrives.indexOf(drive.dev);
            if (index > -1) {
                return update(selectedDrives, { $splice: [[index, 1]] });
            }
            return update(selectedDrives, { $push: [drive.dev] });
        });
    }
    return (
        <tr>
            <td className={drive.dev.length > 3 ? "text-center" : "".trim()}>
                <CheckBox
                    checked={isActiveAndSelected || isSelected}
                    label={drive.dev}
                    onChange={() =>
                        isActiveAndSelected
                            ? setIsActiveAndSelected(false)
                            : select()
                    }
                    disabled={disabled}
                />
            </td>
            <td>{drive.description}</td>
            <td>{drive.fs}</td>
            <td>{drive.uuid}</td>
        </tr>
    );
}
