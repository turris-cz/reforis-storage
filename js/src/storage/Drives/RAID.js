/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import { Select } from "foris";

export const RAID_CHOICES = {
    custom: _("Custom"),
    single: _("JBOD"),
    raid1: _("RAID1"),
};

RAIDSelect.propTypes = {
    selectedRAID: PropTypes.string.isRequired,
    setSelectedRAID: PropTypes.func.isRequired,
    storageIsPending: PropTypes.bool.isRequired,
};

const RAID_HELP_TEXTS = {
    custom: _(`
Don't change the RAID level, keeps everything set the way it was. Useful if you have a really custom setup of your RAID
we don't support and want to just add/remove some disks.
`),
    JBOD: _(`
No redundancy, if one drive fails, you loose your data. On the other hand, provides the most space - sum of the space
available on all included drives.
`),
    RAID1: _(`
Everything is redundant and stored at two distinct drives. If you loose one drive, you can easily replace it without
loosing any data. On the other hand, you have a half of the disk space you would have with JBOD.
`),
};

export function RAIDSelect({
    selectedRAID,
    setSelectedRAID,
    storageIsPending,
}) {
    let raidHelpText;
    switch (selectedRAID) {
        case "raid1":
            raidHelpText = `RAID1: ${RAID_HELP_TEXTS.RAID1}`;
            break;
        case "single":
            raidHelpText = `JBOD: ${RAID_HELP_TEXTS.JBOD}`;
            break;
        default:
            raidHelpText = `Custom: ${RAID_HELP_TEXTS.custom}`;
            break;
    }

    return (
        <Select
            label={_("RAID")}
            choices={RAID_CHOICES}
            value={selectedRAID}
            onChange={(e) => {
                setSelectedRAID(e.target.value);
            }}
            disabled={storageIsPending}
            helpText={raidHelpText}
        />
    );
}
