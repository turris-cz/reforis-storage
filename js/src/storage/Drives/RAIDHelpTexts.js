/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

const RAID_HELP_TEXTS = {
    custom: _(`
Don't change the RAID level, keeps everything set the way it was. Useful if you have a really custom setup of your RAID
we don't support and want to just add/remove some disks.
`),
    JBOD: _(`
No redundancy, if one drive fails, you loose your data. On the other hand, provides the most space - sum of the space
available on all included drives.
`),
    RAID1: _(`
Everything is redundant and stored at two distinct drives. If you loose one drive, you can easily replace it without
loosing any data. On the other hand, you have a half of the disk space you would have with JBOD.
`),
};

export default function RAIDHelpTexts() {
    return (
        <div className="container-fluid">
            <div className="row row-cols-3">
                <div className="col-12 col-lg-4">
                    <h3>{_("Custom")}</h3>
                    <p>{RAID_HELP_TEXTS.custom}</p>
                </div>
                <div className="col-12 col-lg-4">
                    <h3>{_("JBOD")}</h3>
                    <p>{RAID_HELP_TEXTS.JBOD}</p>
                </div>
                <div className="col-12 col-lg-4">
                    <h3>{_("RAID1")}</h3>
                    <p>{RAID_HELP_TEXTS.RAID1}</p>
                </div>
            </div>
        </div>
    );
}
