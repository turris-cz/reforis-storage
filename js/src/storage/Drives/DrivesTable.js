/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { formFieldsSize } from "foris";
import { DriveTableRow } from "./DrivesTableRow";

export default function DrivesTable({
    drives, selectedDrives, setSelectedDrives, currentUUID, storageIsPending,
}) {
    return (
        <table className={`table table-hover ${formFieldsSize}`}>
            <thead>
                <tr>
                    <th scope="col">{_("Device")}</th>
                    <th scope="col">{_("Description")}</th>
                    <th scope="col">{_("Filesystem")}</th>
                    <th scope="col">{_("UUID")}</th>
                </tr>
            </thead>
            <tbody>
                {drives.map((drive) => (
                    <DriveTableRow
                        key={drive.dev}
                        drive={drive}
                        isSelected={selectedDrives.includes(drive.dev)}
                        setSelectedDrives={setSelectedDrives}
                        active={currentUUID && currentUUID === drive.uuid}
                        disabled={storageIsPending}
                    />
                ))}
            </tbody>
        </table>
    );
}
