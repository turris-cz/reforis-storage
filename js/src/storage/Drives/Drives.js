/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useState } from "react";
import {
    API_STATE, Button, Select, useAlert, useAPIPost,
} from "foris";

import DrivesTable from "./DrivesTable";
import API_URLs from "../../API";

const RAID_CHOICES = {
    custom: _("Custom"),
    single: _("JBOD"),
    raid1: _("RAID1"),
};

export default function Drives({ drives, currentUUID, storageIsPending }) {
    const [selectedDrives, setSelectedDrives] = useState([]);
    const [selectedRAID, setSelectedRAID] = useState(Object.keys(RAID_CHOICES)[0]);

    // TODO: Add modal warning.
    // The modal message is depend if uuid is selected or not.
    const [prepareSrvRequestStatus, prepareSrvRequest] = useAPIPost(API_URLs.prepareSrv);
    const [setAlert] = useAlert();
    useEffect(() => {
        if (prepareSrvRequestStatus.state === API_STATE.ERROR) {
            setAlert(_("Device preparing was failed."));
        }
    }, [prepareSrvRequestStatus.state, setAlert]);

    function prepareSrv(event) {
        event.preventDefault();
        prepareSrvRequest({
            drives: selectedDrives,
            raid: selectedRAID,
        });
    }

    const buttonIsDisabled = (
        selectedDrives.length === 0
        || (selectedDrives.length < 2 && selectedRAID === "raid1")
        || storageIsPending
    );

    return (
        <form onSubmit={prepareSrv}>
            <RAIDSelect
                selectedRAID={selectedRAID}
                setSelectedRAID={setSelectedRAID}
                storageIsPending={storageIsPending}
            />
            <DrivesTable
                drives={drives}
                selectedDrives={selectedDrives}
                setSelectedDrives={setSelectedDrives}
                currentUUID={currentUUID}
                storageIsPending={storageIsPending}
            />
            <Button
                type="submit"
                forisFormSize
                disabled={buttonIsDisabled}
            >
                {_("Format&Set")}
            </Button>
        </form>
    );
}

const RAID_HELP_TEXTS = {
    custom: _(`
Don't change the RAID level, keeps everything set the way it was. Useful if you have a really custom setup of your RAID
we don't support and want to just add/remove some disks.
    `),
    JBOD: _(`
No redundancy, if one drive fails, you loose your data. On the other hand, provides the most space - sum of the space
available on all included drives.
    `),
    RAID1: _(`
Everything is redundant and stored at two distinct drives. If you loose one drive, you can easily replace it without
loosing any data. On the other hand, you have a half of the disk space you would have with JBOD.
        `),
};

function RAIDHelpTexts() {
    return (
        <>
            <h5>{_("Custom")}</h5>
            <p>{RAID_HELP_TEXTS.custom}</p>
            <h5>{_("JBOD")}</h5>
            <p>{RAID_HELP_TEXTS.JBOD}</p>
            <h5>{_("RAID1")}</h5>
            <p>{RAID_HELP_TEXTS.RAID1}</p>
        </>
    );
}

function RAIDSelect({ selectedRAID, setSelectedRAID, storageIsPending }) {
    return (
        <>
            <RAIDHelpTexts />
            <Select
                label={_("RAID")}
                choices={RAID_CHOICES}
                value={selectedRAID}
                onChange={(e) => {
                    setSelectedRAID(e.target.value);
                }}

                disabled={storageIsPending}
            />
        </>
    );
}
