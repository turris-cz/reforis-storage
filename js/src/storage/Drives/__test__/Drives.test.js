/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render,
    wait,
    fireEvent,
    getByText,
    getByLabelText,
} from "foris/testUtils/customTestRender";
import mockAxios from "jest-mock-axios";

import diffSnapshot from "snapshot-diff";
import { mockJSONError } from "foris/testUtils/network";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import drives from "../../__tests__/__fixtures__/drives";
import Drives from "../Drives";

describe("<Drives />", () => {
    let container;
    let asFragment;

    beforeEach(() => {
        ({ container, asFragment } = render(
            <>
                <div id="modal-container" />
                <Drives currentUUID="" storageIsPending={false} {...drives} />
            </>
        ));
    });

    it("Should render.", () => {
        expect(container).toMatchSnapshot();
    });

    it("Button should be disabled when nothing is selected.", () => {
        expect(getByText(container, "Format & Set").disabled).toBeTruthy();
    });

    it("Button should be enabled when some disk is selected.", () => {
        fireEvent.click(getByLabelText(container, "sdb"));
        expect(getByText(container, "Format & Set").disabled).toBeFalsy();
    });

    it("Button should be disabled when RAID1 and less that two disks are selected.", () => {
        fireEvent.change(getByLabelText(container, "RAID"), {
            target: { value: "raid1" },
        });
        fireEvent.click(getByLabelText(container, "sdb"));
        expect(getByText(container, "Format & Set").disabled).toBeTruthy();

        fireEvent.click(getByLabelText(container, "sdc"));
        expect(getByText(container, "Format & Set").disabled).toBeFalsy();
    });

    it("Should render modal on Format&Set click.", () => {
        fireEvent.change(getByLabelText(container, "RAID"), {
            target: { value: "single" },
        });
        fireEvent.click(getByLabelText(container, "sdb"));
        fireEvent.click(getByLabelText(container, "sdc"));
        const firstRender = asFragment();

        fireEvent.click(getByText(container, "Format & Set"));
        getByText(container, "Continue");

        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it("Should post on 'Continue' button click.", () => {
        fireEvent.change(getByLabelText(container, "RAID"), {
            target: { value: "single" },
        });
        fireEvent.click(getByLabelText(container, "sdb"));
        fireEvent.click(getByLabelText(container, "sdc"));
        fireEvent.click(getByText(container, "Format & Set"));
        fireEvent.click(getByText(container, "Continue"));

        const data = {
            drives: ["sdb", "sdc"],
            raid: "single",
        };
        expect(mockAxios.post).toHaveBeenCalledWith(
            "/reforis/storage/api/prepare-srv",
            data,
            expect.anything()
        );
    });

    it("Should handle post error.", async () => {
        fireEvent.click(getByLabelText(container, "sdb"));
        fireEvent.click(getByText(container, "Format & Set"));
        fireEvent.click(getByText(container, "Continue"));
        mockJSONError();
        await wait(() =>
            expect(mockSetAlert).toBeCalledWith("Device preparation failed.")
        );
    });
});
