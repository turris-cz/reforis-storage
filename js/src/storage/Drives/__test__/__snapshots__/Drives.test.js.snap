// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drives /> Should render modal on Format&Set click. 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -3,11 +3,11 @@
      id=\\"modal-container\\"
    >
      <div
        aria-labelledby=\\"modal-title\\"
        aria-modal=\\"true\\"
-       class=\\"modal fade\\"
+       class=\\"modal fade show\\"
        role=\\"dialog\\"
      >
        <div
          class=\\"modal-dialog modal-dialog-centered\\"
          role=\\"document\\""
`;

exports[`<Drives /> Should render. 1`] = `
<div>
  <div
    id="modal-container"
  />
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="1"
    >
      RAID
    </label>
    <select
      class="form-select"
      id="1"
    >
      <option
        value="custom"
      >
        Not specified
      </option>
      <option
        value="single"
      >
        JBOD
      </option>
      <option
        value="raid1"
      >
        RAID1
      </option>
    </select>
    <div
      class="form-text"
    >
      <small>
        Not specified: 
Don't change the RAID level, keeps everything set the way it was. Useful if you have a really custom setup of your RAID
we don't support and want to just add/remove some disks.

      </small>
    </div>
  </div>
  <div
    class="table-responsive"
  >
    <table
      class="table table-hover"
    >
      <thead
        class="thead-light"
      >
        <tr>
          <th
            scope="col"
          >
            Device
          </th>
          <th
            scope="col"
          >
            Description
          </th>
          <th
            scope="col"
          >
            Filesystem
          </th>
          <th
            scope="col"
          >
            UUID
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div
              class="mb-0 form-check"
            >
              <input
                class="form-check-input"
                id="2"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="2"
              >
                sdb
              </label>
            </div>
          </td>
          <td>
            USB DISK 2.0 (7.2 GiB)
          </td>
          <td />
          <td>
            1111
          </td>
        </tr>
        <tr>
          <td>
            <div
              class="mb-0 ms-4 form-check"
            >
              <input
                class="form-check-input"
                id="3"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="3"
              >
                sdb1
              </label>
            </div>
          </td>
          <td>
            USB DISK (7.2 GiB)
          </td>
          <td>
            ext4
          </td>
          <td>
            3333
          </td>
        </tr>
        <tr>
          <td>
            <div
              class="mb-0 form-check"
            >
              <input
                class="form-check-input"
                id="4"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="4"
              >
                sdc
              </label>
            </div>
          </td>
          <td>
            USB DISK 2.0 (3.6 GiB)
          </td>
          <td>
            btrfs
          </td>
          <td>
            4444
          </td>
        </tr>
        <tr>
          <td>
            <div
              class="mb-0 ms-4 form-check"
            >
              <input
                class="form-check-input"
                id="5"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="5"
              >
                sdc1
              </label>
            </div>
          </td>
          <td>
            USB DISK (3.6 GiB)
          </td>
          <td>
            vfat
          </td>
          <td>
            2222
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="text-end"
  >
    <button
      class="btn btn-primary mb-2 mb-md-0 mb-lg-3 ml-auto col-12 col-md-4 col-lg-2 d-inline-flex justify-content-center align-items-center"
      data-testid="format-set-btn"
      disabled=""
      type="button"
    >
      <span>
        Format & Set
      </span>
    </button>
  </div>
</div>
`;
